include/odp/api/spec/ipsec.h

```diff
@@ -902,6 +907,9 @@ typedef struct odp_ipsec_error_t {
 
 			/** Hard lifetime expired: packets */
 			uint32_t hard_exp_packets : 1;
+
+			/** TFC dummy packet (NextHeader = 59) */
+			uint32_t tfc_dummy_packet : 1;
```

Comment:
I suggest that when implementation does not drop "no next header" packets, those come through as any other successfully transformed packets: no error/warning bits set. However, when HW has dropped a no NH packet it generates a dummy packet (length zero, etc) with a warning or operation flag bit on telling that this is not a normal packet, but an indication that a no NH packet was dropped.

Errors/warnings indicate a failure and may require some specific action (SA negotiation). This instead could be could be just a odp_ipsec_op_flag_t bit, as the operation was successful and does not need specific action (other than drop packet after finding out that it does not contain any data). So, application could ignore the bit and drop packet as part of normal processing of too short packets.

Reply:
They can not come as "normal" packets. An application will have no way to determine that the package was dummy after receiving it. NH field will be gone in case of tunnel packets.

Reply:
I prefer to have this bit as an error to keep very simple constraint. No error => app may forward packet (if it is not interested in details). Error => packet must be threated separately and must not be forwarded.

Reply:
A dummy packet in transport mode could be delivered to an application in the normal way. The result packet would be an IP packet with protocol 59. But it is true that in tunnel mode there is no way to deliver the packet to the application in the current API. In my interpretation of the current API a tunnel mode dummy packet would result in a proto error (since the decapsulated packet is neither IPv4 nor IPv6).

So some warning/error/flag bit is indeed needed if one does not want to confuse dummy packets with protocol errors.

Reply:
Currently the API promises to give the original packet back to the application in case of any error. This would apply to dummy packets too. Maybe the API should be so that such preservation of the original packet is not required in the case of dummy packets to reduce the overhead of handling dummy packets.

Reporting dummy packet through an error is a bit misleading since no error happened. An application cannot e.g. have a global IPsec error counter without first filtering out dummy packets.

Reply:
Maybe a generic solution would be to add the IPSEC next header field value into odp_ipsec_packet_result_t. I guess all implementations should be able to return that after decrypt as RFCs mention also TCP as one possible value. So, payload could be at least IPv4, IPv6, TCP or NONH.

" The value of this
   field is chosen from the set of IP Protocol Numbers defined on the
   web page of Internet Assigned Numbers Authority (IANA).  For example,
   a value of 4 indicates IPv4, a value of 41 indicates IPv6, and a
   value of 6 indicates TCP."

Reply:
I do not think it is necessary. The next header field can be anything (like TCP) in transport mode, but in transport mode this API delivers the application and IP packet and inserts the protocol value from the next header field in the protocol field of the IP header. For tunnel mode SAs other next header values than those for IPv4, IPv6 and dummy packets should not appear.

Reply:
The issue with transport mode ESP and NONH is that we should never forward NOHN packets even in transport mode. So both transport and tunnel NONH packets should raise same exceptions.

Reply:
I was just pointing out that from functional completeness perspective API changes are not strictly needed. But as was discussed in the Wednesday meeting, having the next header field in the packet result could be useful and convenient anyway.

Reply:
In case if we update the warning bit when the dummy packet is found then there will not be much impact on the application side since odp_ipsec_status_t consists both error and warning bits. Updating next proto field in ipsec_op_packet_result_t requires implementation to process and update this field for every packet.

The only API change required is to specify clearly that in-case of a dummy packet the implementation will not do any selector checks on the received packets.

Reply:
What selector checks are you referring to? Traffic selectors are handled fully in the application side. The ODP implementation does not know the traffic selectors and cannot do any checks anyway.

Reply:
I was referring to ODP_IPSEC_LOOKUP_DSTADDR_SPI. This cannot be matched in case of NH   = 59.  Actually, this raises another question. Do you really need the SA from which this Dummy packet was received in the application? Like do we have to fill the SA value in the odp_ipsec_packet_result_t field?

Reply:
ODP_IPSEC_LOOKUP_DSTADDR_SPI means that for an SA to match a received packet, both the SPI and the destination IP address of the received ESP/AH packet must match those configured in the SA. This lookup must happen before IPsec processing (decryption), i.e. before it is even possible to know that the packet is actually a dummy TFC packet. You have to find the correct SA and perform decryption before you can tell whether a received packet is a genuine data packet or a dummy TFC packet.

An application expects the used SA to be set in the packet result and, IMHO, dummy packets should not be any different since they got processed using the SA.

Reply:
Hmm... I misread the specification so LOOKUP_DSTADDR_SPI only uses ip addr and proto field for lookup and does not validate the src/dst ip address or proto filed after decryption. I believe this validation of IP addr and version can be done by the HW and could be something to add as an enhancement to the spec in the future.

Yes, we could add this SA to packet result but since the RFC specifies in the implementation note to discard the dummy packet after the reception I was inclined towards not updating these structures for dummy packets.

What is your opinion on updating the next protocol field in odp_packet_result vs adding a warning bit for dummy packet?

Reply:
We might add next_proto field to packet_result. @NikhilA-Linaro how hard would it be to get this field on your platform?

I'm against using warning bit for dummy packets. From ODP packet flow point of view, dummy packet is clearly an errorneous packet.

Reply:
From IPSEC operation point of view a successfully decrypted dummy packet is a successful operation (SA matched, hash was OK, decrypt was OK, protocol headers were OK, no replay issues, etc). Just the payload format is different than usual. If TFC is really used to hide traffic pattern, all spare capacity of the connection could be filled by dummy packets. So, it would be more common for application to receive these vs. the normal packet (as long as HW is not dropping dummies).

So, I'd like to reserve error/warning bits for the cases when there's really a problem (SA configuration miss-match, attack, broken sender implementation, etc).

Reply:
@psavol I agree with your arguments. However I'm worried about fast-forwarding use case. Do you think it will make sense to make tfc-packet warning (or even just flag) but still require that `odp_packet_has_error()` returns true for such packets?

Reply:
Application needs to anyway check if
 * packet/ipsec operation result has errors
 * IP version of packet (== what's in the tunnel)

If we add result.next_header field, application can use that for both checking IP version and NONH in one switch case. There's no extra check for NONH.

    if (result.status != 0)
        goto error;

    switch (result.next_header) {
        case IPV4:
            ....
        case IPV6:
            ....
        case NONH:
            odp_packet_free(pkt);
            break;
        default:
            goto error;

Reply:
@psavol Yes, but it requires to call `odp_ipsec_result()` (which can be quite heavyweight).  Currently it is possible to forward packets right after calling `odp_packet_has_error()` and `odp_packet_has_ipv4()`, which can be inlined to just bit checks.

Anyway, this is now a question to @bala-manoharan and @NikhilA-Linaro if it would be easy to add next_header field to `odp_ipsec_packet_result_t`

Reply:
@bala-manoharan @NikhilA-Linaro Please provide input here.